File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        
        List<String> answer = new ArrayList<>();
        StringBuilder sb = new StringBuilder("");
        helper(answer, 2*n, sb, 0, 0);
        return answer;
    }

    public void helper(List<String> answer, int n, StringBuilder sb, int index, int sum) {
        //base case
        if(index >= n) {
            if(sum == 0){
                answer.add(sb.toString());
            }
            return;
        }

        //recursive case
        if(sum < n/2) {
            sb.append('(');
            helper(answer, n, sb, index + 1, sum + 1);
            sb.deleteCharAt(sb.length()-1);
        }

        if(sum > 0) {
            sb.append(')');
            helper(answer, n, sb, index + 1, sum - 1);
            sb.deleteCharAt(sb.length()-1);
        }
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:

    vector<string> answer = {};
    vector<char> currentSeq = {};

    vector<string> generateParenthesis(int n) {

        recursion(0, 0, 2*n);
        return answer;
    }

    void recursion(int start, int sum, int L) {

        //base case
        if(start == L) {
            if(sum == 0) {

                string str = "";
                for(int i=0; i<currentSeq.size(); i++) {
                    str += currentSeq[i];
                }
                answer.push_back(str);
            }
            return;
        }

        //recursive case
        //1st for open paranthesis
        if(sum < L/2) {
            currentSeq.push_back('(');
            recursion(start+1, sum+1, L);
            currentSeq.pop_back();
        }
        

        //2nd for closing paranthesis
        if(sum > 0) {
            currentSeq.push_back(')');
            recursion(start+1, sum-1, L);
            currentSeq.pop_back();
        }

    }
};